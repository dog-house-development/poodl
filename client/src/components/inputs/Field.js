import React, { Component } from "react";
import PropTypes from "prop-types";

const propTypes = {
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string,
  id: PropTypes.string.isRequired,
  type: PropTypes.string,
  size: PropTypes.oneOf(["normal", "large"]),
  content: PropTypes.string,
  placeholder: PropTypes.string,
  label: PropTypes.string,
  error: PropTypes.object
};

const defaultProps = {
  size: "normal",
  type: "text"
};

class Field extends Component {
  constructor(props) {
    super(props);
    this.state = { value: "" };
    this.handleChange = this.handleChange.bind(this);
  }

  onComponentDidMount() {
    this.setState({ value: this.props.content });
  }

  handleChange(e) {
    e.preventDefault();
    this.setState({ value: e.target.value });
    this.props.onChange(e);
  }

  render() {
    return (
      <div>
        {this.props.label}
        <input
          id={this.props.id}
          type={this.props.type}
          name={this.props.name}
          value={this.state.value}
          placeholder={this.props.placeholder}
          onChange={this.handleChange}
          className={`field ${this.props.size}`}
          error={this.props.error}
        />
      </div>
    );
  }
}

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;
export default Field;
