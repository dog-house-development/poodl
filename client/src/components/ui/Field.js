import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _ from 'lodash';

const propTypes = {
    onChange: PropTypes.func.isRequired,
    name: PropTypes.string,
    id: PropTypes.string.isRequired,
    type: PropTypes.string,
    min: PropTypes.string,
    max: PropTypes.string,
    size: PropTypes.oneOf(['normal', 'large']),
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    kind: PropTypes.oneOf(['primary', 'secondary']),
    placeholder: PropTypes.string,
    label: PropTypes.string,
    autoComplete: PropTypes.oneOf(['on', 'off']),
    autoFocus: PropTypes.oneOf(['on', 'off']),
    spellCheck: PropTypes.oneOf(['true', 'false']),
    sidebyside: PropTypes.oneOf([1, 2]),
    error: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    onClick: PropTypes.func,
    present: PropTypes.bool,
    clearable: PropTypes.bool,
    leftIcon: PropTypes.string
};

const defaultProps = {
    size: 'normal',
    type: 'text',
    autoComplete: 'on',
    value: '',
    kind: 'primary'
};

class Field extends Component {
    constructor(props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange(e) {
        e.preventDefault();
        this.props.onChange(e);
    }

    getClearMarkup() {
        if (this.props.clearable && !_.isEmpty(this.props.value)) {
            return (
                <i
                    className={classnames('material-icons', 'clear-icon', this.props.size, this.props.kind)}
                    onClick={this.onClearClick}>
                    clear
                </i>
            );
        }
    }

    onClearClick = () => {
        const e = { target: { value: '' } };
        this.props.onChange(e);
    };

    getLeftIconMarkup() {
        if (this.props.leftIcon) {
            return <i className={classnames('material-icons', 'left-icon', this.props.size)}>{this.props.leftIcon}</i>;
        }
    }

    render() {
        const { style, present, editable, clearable, leftIcon, ...inputProps } = this.props;
        if (present) {
            return (
                <div style={style} className="field-wrapper editable-field-wrapper">
                    <p className="field-label">{this.props.label}</p>
                    <p>{this.props.value}</p>
                </div>
            );
        }

        return (
            <div
                style={style}
                className={classnames('field-wrapper', this.props.className, {
                    'inline-field': this.props.sidebyside
                })}>
                <p className="field-label">{this.props.label}</p>
                <div className="field-outer">
                    {this.getLeftIconMarkup()}
                    <input
                        {...inputProps}
                        className={classnames(
                            'field',
                            this.props.size,
                            this.props.kind,
                            { 'first-side-by-side-input': this.props.sidebyside === 1 },
                            { 'second-side-by-side-input': this.props.sidebyside === 2 },
                            { 'field-error-border': this.props.error },
                            { 'clearable-field': this.props.clearable },
                            { 'include-left-icon': this.props.leftIcon }
                        )}
                    />
                    {this.getClearMarkup()}
                    <p className="field-error-label">{this.props.error}</p>
                </div>
            </div>
        );
    }
}

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;
export default Field;
