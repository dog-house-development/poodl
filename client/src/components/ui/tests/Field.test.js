import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import Field from '../Field';

configure({ adapter: new Adapter() });

describe('Field tests', () => {
    let wrapper, instance;
    const setInstanceAndWrapper = (_props = {}) => {
        const props = _.assign(
            {},
            {
                onChange: () => {},
                name: 'email',
                id: 'email',
                type: 'email',
                size: 'normal',
                value: '',
                content: '',
                placeholder: 'Enter email...',
                label: 'Email:',
                error: {}
            },
            _props
        );
        wrapper = shallow(<Field {..._.assign({}, props)} />);
        instance = wrapper.instance();
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('handleChange', () => {
        it('should change set the state to the new value', () => {
            spyOn(instance, 'handleChange');
            const baseProps = { onChange: jest.fn() };
            setInstanceAndWrapper(baseProps);
            const e = {
                target: { name: 'email', value: 'Moo' },
                preventDefault: () => {}
            };
            wrapper.find('input').simulate('change', e);
            expect(baseProps.onChange).toHaveBeenCalledTimes(1);
        });
    });

    describe('render', () => {
        it('should render not side by side', () => {
            setInstanceAndWrapper();
            expect(wrapper).toMatchSnapshot();
        });

        it('should render the first side by side input', () => {
            setInstanceAndWrapper({ sidebyside: 1 });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render the second side by side input', () => {
            setInstanceAndWrapper({ sidebyside: 2 });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render without errors correctly', () => {
            setInstanceAndWrapper({ error: null });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render with errors correctly', () => {
            setInstanceAndWrapper({ error: 'Big bad error' });
            expect(wrapper).toMatchSnapshot();
        });
    });
});
