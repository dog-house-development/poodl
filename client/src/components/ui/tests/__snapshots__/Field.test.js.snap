// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field tests render should render not side by side 1`] = `
<div
  className="field-wrapper"
>
  <p
    className="field-label"
  >
    Email:
  </p>
  <div
    className="field-outer"
  >
    <input
      autoComplete="on"
      className="field normal primary field-error-border"
      content=""
      error={Object {}}
      id="email"
      kind="primary"
      label="Email:"
      name="email"
      onChange={[Function]}
      placeholder="Enter email..."
      size="normal"
      type="email"
      value=""
    />
    <p
      className="field-error-label"
    >
      <Component />
    </p>
  </div>
</div>
`;

exports[`Field tests render should render the first side by side input 1`] = `
<div
  className="field-wrapper inline-field"
>
  <p
    className="field-label"
  >
    Email:
  </p>
  <div
    className="field-outer"
  >
    <input
      autoComplete="on"
      className="field normal primary first-side-by-side-input field-error-border"
      content=""
      error={Object {}}
      id="email"
      kind="primary"
      label="Email:"
      name="email"
      onChange={[Function]}
      placeholder="Enter email..."
      sidebyside={1}
      size="normal"
      type="email"
      value=""
    />
    <p
      className="field-error-label"
    >
      <Component />
    </p>
  </div>
</div>
`;

exports[`Field tests render should render the second side by side input 1`] = `
<div
  className="field-wrapper inline-field"
>
  <p
    className="field-label"
  >
    Email:
  </p>
  <div
    className="field-outer"
  >
    <input
      autoComplete="on"
      className="field normal primary second-side-by-side-input field-error-border"
      content=""
      error={Object {}}
      id="email"
      kind="primary"
      label="Email:"
      name="email"
      onChange={[Function]}
      placeholder="Enter email..."
      sidebyside={2}
      size="normal"
      type="email"
      value=""
    />
    <p
      className="field-error-label"
    >
      <Component />
    </p>
  </div>
</div>
`;

exports[`Field tests render should render with errors correctly 1`] = `
<div
  className="field-wrapper"
>
  <p
    className="field-label"
  >
    Email:
  </p>
  <div
    className="field-outer"
  >
    <input
      autoComplete="on"
      className="field normal primary field-error-border"
      content=""
      error="Big bad error"
      id="email"
      kind="primary"
      label="Email:"
      name="email"
      onChange={[Function]}
      placeholder="Enter email..."
      size="normal"
      type="email"
      value=""
    />
    <p
      className="field-error-label"
    >
      Big bad error
    </p>
  </div>
</div>
`;

exports[`Field tests render should render without errors correctly 1`] = `
<div
  className="field-wrapper"
>
  <p
    className="field-label"
  >
    Email:
  </p>
  <div
    className="field-outer"
  >
    <input
      autoComplete="on"
      className="field normal primary"
      content=""
      error={null}
      id="email"
      kind="primary"
      label="Email:"
      name="email"
      onChange={[Function]}
      placeholder="Enter email..."
      size="normal"
      type="email"
      value=""
    />
    <p
      className="field-error-label"
    />
  </div>
</div>
`;
